source/common/thread_local/thread_local_impl.cc

```diff
@@ -17,93 +17,121 @@ thread_local InstanceImpl::ThreadLocalData InstanceImpl::thread_local_data_;
   ASSERT(idx < slots_.size());
   SlotImplPtr slot(new SlotImpl(*this, idx));
   slots_[idx] = slot.get();
-  return std::make_unique<Bookkeeper>(*this, std::move(slot));
+  return slot;
+}
+
+InstanceImpl::SlotImpl::SlotImpl(InstanceImpl& parent, uint64_t index)
+    : parent_(parent), index_(index), still_alive_guard_(std::make_shared<bool>()),
+      active_callback_guard_(
+          new bool,
+          [&parent, still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), this](bool* b) {
+            // If TLS is in the running state, signal that this slot is ready to be destroyed. This
+            // may happen on a worker when callbacks are run. We do not do this check during:
+            // a) initializing: there are error flows during initialization in which it is too
+            //    complicated to destroy all pending callbacks before the slot is destroyed. Given
+            //    that no workers are running there is nothing to synchronize anyway.
+            // b) shutting down: no slots will be destroyed between the time shutdown starts and all
+            //    workers exit, so there is nothing to synchronize once all workers have exited.
+            //    Additionally, there are initialization failure cases in which we move directly
+            //    from initializing to shutting down, so for similar reasons to (a) we don't
+            //    synchronize.
+            // c) when we can't lock the still alive guard. It's only possible for that to happen
+            //    when a slot was destroyed before we entered the running state (during xDS init).
+            //    This is guaranteed by the blocking in slot destruction. Thus, no further
+            //    synchronization is needed.
+            if (parent.state_ == State::Running && still_alive_guard.lock()) {
+              absl::MutexLock lock(&shutdown_mutex_);
+              ready_to_destroy_ = true;
+            }
+
+            delete b;
+          }) {}
+
+InstanceImpl::SlotImpl::~SlotImpl() {
+  // Reset active_callback_guard_ so the only active references will be held by pending callbacks.
+  active_callback_guard_.reset();
+  // If running only, synchronize slot removal with all callbacks being drained from workers.
+  // See the constructor for why we only do this in the running state.
+  if (parent_.state_ == State::Running) {
+    absl::MutexLock lock(&shutdown_mutex_);
+    shutdown_mutex_.Await(absl::Condition(&ready_to_destroy_));
+  }
+
+  parent_.removeSlot(*this);
+}
+
+Event::PostCb InstanceImpl::SlotImpl::wrapCallback(Event::PostCb cb) {
+  // Capture active_callback_guard_ so that we can synchronize slot destruction. See ~SlotImpl().
+  // Also capture still_alive_guard_ so we can verify the slot is still alive as it might have
+  // been destroyed pre-running state. See SlotImpl().
+  return [active_callback_guard = active_callback_guard_,
+          still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), cb] {
```

Comment:
Design alternative: what if this was a strong_ptr (or just use a refcnt at that point), and always allow the `cb` to run, to simplify reasoning about tls callbacks?

Reply:
Can you explain more? I don't follow.

Reply:
Nevermind